wallet: Move wallets loading out from the main GUI thread

hebasto opened the pull request:
This PR improves the GUI responsiveness during initial wallets loading at startup (especially ones that have tons of txs), and shows a standard progress dialog for long loading:

![DeepinScreenshot_select-area_20210522230626](https://user-images.githubusercontent.com/32963518/119239625-0b3a9380-bb53-11eb-9a54-34980d8a1194.png)

Fixes #247.

jarolrod commented:
Some occurrences of a seg fault in tests:
- https://bitcoinbuilds.org/index.php?ansilog=327adf0f-3a42-450e-9f97-21c1c6d556f6.log#l2079
- https://cirrus-ci.com/task/6701673100869632?logs=ci#L4000

promag commented on src/qt/walletcontroller.cpp:
```diff
@@ -190,6 +190,7 @@ WalletControllerActivity::WalletControllerActivity(WalletController* wallet_cont
     , m_wallet_controller(wallet_controller)
     , m_parent_widget(parent_widget)
 {
+    connect(this, &WalletControllerActivity::finished, this, &QObject::deleteLater);
```
4517c51b8605fbf91e2319b16d7d5dcf80cb1565

Originally the idea was to allow the user to see activity details and then the user could dismiss or it could auto dismiss. I also think this is an unrelated change - this PR could also connect `LoadWalletsActivity` finished. But this is also fine and can be improved if later the original idea is implemented.

promag reviewed (commented):
Concept ACK.

ryanofsky reviewed (approved):
Light code review ACK 8c9a9a39876afc650ebb3a54becd652e9b03e66f

I do think it is unfortunate so much boilerplate (WalletControllerActivity class definition, multiple methods, multiple single shot timers and signals) is added just to run 3 lines of code asynchronously.

It should be possible to have reusable utility:

```c++
RunAsynchronously([this] {
   for (std::unique_ptr<interfaces::Wallet>& wallet : node().walletClient().getWallets()) {
       m_wallet_controller->getOrCreateWallet(std::move(wallet));
   }
})
```

I did implement something like this previously (but with more features) as an experiment in https://github.com/bitcoin/bitcoin/compare/master...ryanofsky:pr/gsync comment https://github.com/bitcoin/bitcoin/issues/16874#issuecomment-632424392

promag commented:
@ryanofsky agree, that or `QtConcurrent::run` or `QMetaObject::invokeMethod` (requires min Qt 5.10 IIRC) to just run some code async. But if you want to show progress feedback, update something with "loading wallet x", allow the user to interrupt, maybe pause, at the end ask something else, then wrapping the activity state in some class makes sense to me.

promag reviewed (commented):
@hebasto what do you think about splitting this PR?

hebasto commented:
@promag
> @hebasto what do you think about splitting this PR?

Sounds good. In what parts?

ShaMan239 reviewed (commented):
tACK 8c9a9a39876afc650ebb3a54becd652e9b03e66f
Tested on Ubuntu 20.04 (Using Qt version 5.12.8)

This PR adds a ‘Loading Wallet’ progress window for the initial loading of opened wallet after opening GUI. This PR also moves wallet loading out of the main GUI thread while simplifying some other aspects of the codebase.

I was able to test the PR successfully. I have added a screenshot for the resultant progress window I observed while opening the GUI.
![Screenshot from 2021-09-06 20-06-46](https://user-images.githubusercontent.com/85434418/132235945-85fefd7a-a7c6-4459-bbc3-c89af50c6f58.png)

The changes made in this PR are pretty extensive and detailed. Thanks for this fantastic work, @hebasto.
It took me quite a while to understand everything that has been done in this PR. That’s why I am sharing my notes that discuss the changes in quite detail. So that other reviewers could have an easier time reviewing this PR. If some of my observations are erroneous, please do inform me.

NOTES:
- **qt, macos: Fix GUIUtil::PolishProgressDialog bug** 
    - Done to prevent misbehaving of this PR in macOS.
- **qt: Improve GUI responsiveness** 
    - Removed `Q_UNUSED` because now **dialog** is being used at least once in any case under this function.
    - `setMinimumDuration` is set through the `PolishProgressDialog` function. Since this function is used multiple times when setting `setMinimumDuration` was required, it makes sense to move that in this function. This change simplifies the code while maintaining the same functionality.
    - Under `WalletControllerActivity::showProgressDialog` function, `setValue` is used for the `m_progress_dialog` variable. This starts the internal duration estimation.
    - The now redundant `setMinimumDuration` instruction from `BitcoinGUI::showProgress` and `WalletView::showProgress` functions.
- **qt, wallet: Move activity progress dialog from data member to local**
    - Since the activity progress dialog has been moved from data member to local variable, subsequent changes have been made:
        - `WalletControllerActivity` destructor has been restored to default destructor.
        - In `WalletControllerActivity::showProgressDialog` function:
            - `m_progress_dialog` variable has been renamed to `progress_dialog` to avoid naming confusion with a data member.
            - The non-existence of `m_progress_dialog` (or progress_dialog) is not being asserted now.
        - The `destroyProgressDialog` function is also removed. Because now progress_dialog doesn’t exist out of the function scope and hence no need for a `progress_dialog` destructor.
- **qt, wallet, refactor: Move connection to QObject::deleteLater to ctor**
    - Instead of creating similar connections multiple times for individual objects of class, the connection has been moved to the constructor of the common parent class itself, resulting in simplification of the code.
- **qt, wallet: Add LoadWalletsActivity**
    - New `LoadWalletsActivity` class is declared as a child class of the `WalletControllerActivity` class.
    - A function `Load` is added to the class that calls the `showProgressDialog` function and loads the wallets.
    - This new class and its functions are appropriately declared under the `walletcontroler.h` file
- **qt, wallet: Move wallets loading out from the main GUI thread** 
    - The new `LoadWalletActivity::Load` function is now used to load wallets at the start of the GUI.
    - The redundant code, which is now part of the Load function, has been removed from the WalletController constructor.

I hope these notes will be of some help to fellow reviewers.
<p>

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -818,7 +818,6 @@ void PolishProgressDialog(QProgressDialog* dialog)
     // Workaround for macOS-only Qt bug; see: QTBUG-65750, QTBUG-70357.
     const int margin = TextWidth(dialog->fontMetrics(), ("X"));
     dialog->resize(dialog->width() + 2 * margin, dialog->height());
-    dialog->show();
```
> By "break" you mean it always shows, doesn't take into account `minimumDuration`.

Exactly.

> Only macos?

The removed code is placed within `#ifdef Q_OS_MAC`.

hebasto commented:
Rebased 8c9a9a39876afc650ebb3a54becd652e9b03e66f -> e07509867ec14ccb4f2f5bce082f73b77ade7743 ([pr342.02](https://github.com/hebasto/gui/commits/pr342.02) -> [pr342.03](https://github.com/hebasto/gui/commits/pr342.03)).

ryanofsky commented on src/qt/walletcontroller.cpp:
```diff
@@ -342,3 +342,21 @@ void OpenWalletActivity::open(const std::string& path)
+    showProgressDialog(tr("Loading wallets…"));
+
+    QTimer::singleShot(0, worker(), [this] {
+        for (std::unique_ptr<interfaces::Wallet>& wallet : node().walletClient().getWallets()) {
+            m_wallet_controller->getOrCreateWallet(std::move(wallet));
+        }
```
In commit "qt, wallet: Add LoadWalletsActivity" (6edff4dfa508100564c3152bbccc7b37ac777528)

Note: This is not new code, this is existing code which is just being moved. The original code is deleted in the next commit. I would squash two commits for more clarity here, but it's not important, and of course the best way to split and group commits is subjective.

ryanofsky reviewed (approved):
Code review ACK e07509867ec14ccb4f2f5bce082f73b77ade7743. I still think the need for these activity classes and timer hacks is unfortunate, because it should be possible to have a utility function that takes a lambda, runs it asynchronously and posts a result back to the GUI thread. (The lambda would also be able to capture variables so would not be precluded from receiving input or posting incremental feedback either.) But the activity class here is pretty small and this suggestion isn't very important, just something to keep in mind if we are looking to reduce boilerplate and improve readability later.

ryanofsky commented:
It seems like this would fix https://github.com/bitcoin-core/gui/issues/247 by making the gui responsive during walletmodel loading. Is this the case? It might also be good to mention in the PR description

hebasto commented:
@ryanofsky @shaavan @promag 

Thank you for your reviews!

Updated e07509867ec14ccb4f2f5bce082f73b77ade7743 -> 2fe69efbc607fdcc3657637d59a38cc5b4db2d05 ([pr342.03](https://github.com/hebasto/gui/commits/pr342.03) -> [pr342.04](https://github.com/hebasto/gui/commits/pr342.04)):

- addressed @ryanofsky [comment](https://github.com/bitcoin-core/gui/pull/342#discussion_r703751149):
> The `getOpenWallets` wallets method seems unused after this change and could be dropped

- addressed @ryanofsky [comment](https://github.com/bitcoin-core/gui/pull/342#discussion_r703759031):
> This is not new code, this is existing code which is just being moved. The original code is deleted in the next commit. I would squash two commits for more clarity here

---
@ryanofsky 
> It seems like this would fix #247 by making the gui responsive during walletmodel loading. Is this the case?

I don't think creating `WalletModel` instances is the most time consuming process. Another part is `WalletClient::start()`. And I believe to fix #247 both PRs are required -- https://github.com/bitcoin/bitcoin/pull/22231 and this one.

> It might also be good to mention in the PR description

Done.

promag commented:
> I don't think creating `WalletModel` instances is the most time consuming process.

This is not true, in fact, it's the most expensive task, because on `WalletModel` constructor `AddressTableModel` and `TransactionTableModel` are also instanced, and these prefetch wallet addresses and transactions.

hebasto commented:
> > I don't think creating `WalletModel` instances is the most time consuming process.
> 
> This is not true, in fact, it's the most expensive task, because on `WalletModel` constructor `AddressTableModel` and `TransactionTableModel` are also instanced, and these prefetch wallet addresses and transactions.

Yeah, `TransactionTablePriv::refreshWallet` could really take much time for big wallets.

hebasto commented:
@ryanofsky 
> It seems like this would fix #247 by making the gui responsive during walletmodel loading. Is this the case? It might also be good to mention in the PR description

I've re-read logs provided by @laanwj in https://github.com/bitcoin-core/gui/issues/247#issue-830828828 and I see now you are right.

promag commented on src/qt/walletcontroller.cpp:
```diff
@@ -356,3 +332,21 @@ void OpenWalletActivity::open(const std::string& path)
+{
+    showProgressDialog(tr("Loading wallets…"));
+
+    QTimer::singleShot(0, worker(), [this] {
```
Use `GUIUtil::ObjectInvoke` here and below?

hebasto replied:
I'd rather keep this PR as is, and in a follow up use `GUIUtil::ObjectInvoke` in a wider scope of all `walletcontroller.cpp`.

promag commented on src/qt/walletcontroller.cpp:
```diff
@@ -356,3 +332,21 @@ void OpenWalletActivity::open(const std::string& path)
+
+    QTimer::singleShot(0, worker(), [this] {
+        for (auto& wallet : node().walletClient().getWallets()) {
+            m_wallet_controller->getOrCreateWallet(std::move(wallet));
```
nit, we could show progress and/or show each wallet name.

ryanofsky reviewed (approved):
Code review ACK 2fe69efbc607fdcc3657637d59a38cc5b4db2d05. Just suggested changed since last review: squashing commits and dropping unused method (thanks!)

shaavan reviewed (approved):
reACK 2fe69efbc607fdcc3657637d59a38cc5b4db2d05

Updates since my last review:

- PR rebased on current master
-  Two Commits. **qt, macOS: Fix GUIUtil::PolishProgressDialog bug** and **qt: Improve GUI responsiveness** are dropped because they are no longer needed.
- Made appropriate changes to other commits and removed parts that are no longer needed or used.
- **qt, wallet: Move wallets loading out from the main GUI thread** and **qt, wallet: Add LoadWalletsActivity class** have been squashed together.
- New commit **qt, wallet: Drop no longer used WalletController::getOpenWallets()** which, as it say so, removed `getOpenWallets()` function.

promag reviewed (commented):
Code review ACK 2fe69efbc607fdcc3657637d59a38cc5b4db2d05.
